File: solutions/zig/yacht/2/yacht.zig
const std = @import("std");
const mem = std.mem;

pub const Category = enum(u4) {
    ones = 1,
    twos,
    threes,
    fours,
    fives,
    sixes,
    full_house,
    four_of_a_kind,
    little_straight,
    big_straight,
    choice,
    yacht,
};

fn count(d: [5]u3, n: u3) u32 {
    const needle = [_]u3{n};
    return @intCast(mem.count(u3, &d, &needle));
}

fn fullHouse(d: [5]u3) bool {
    const f = d[0] == d[1] and d[1] != d[2] and d[2] == d[4]; // AABBB
    const s = d[0] == d[2] and d[2] != d[3] and d[3] == d[4]; // AAABB
    return f or s;
}

fn sum(d: [5]u3) u32 {
    return @as(u32, d[0]) + d[1] + d[2] + d[3] + d[4];
}

pub fn score(dice: [5]u3, category: Category) u32 {
    var sorted: [5]u3 = undefined;
    @memcpy(&sorted, &dice);
    mem.sort(u3, &sorted, {}, comptime std.sort.asc(u3));

    return switch (category) {
        .ones, .twos, .threes, .fours, .fives, .sixes => |c| @intFromEnum(c) * count(sorted, @intCast(@intFromEnum(c))),
        .full_house => if (fullHouse(sorted)) sum(sorted) else 0,
        .four_of_a_kind => if (sorted[0] == sorted[3] or sorted[1] == sorted[4]) @as(u32, 4) * sorted[1] else 0,
        .little_straight => if (mem.eql(u3, &sorted, &.{ 1, 2, 3, 4, 5 })) 30 else 0,
        .big_straight => if (mem.eql(u3, &sorted, &.{ 2, 3, 4, 5, 6 })) 30 else 0,
        .choice => sum(sorted),
        .yacht => if (sorted[0] == sorted[4]) 50 else 0,
    };
}

File: solutions/ruby/yacht/2/yacht.rb
# frozen_string_literal: true

# class Yacht
class Yacht
  attr_reader :score

  def initialize(dice, category)
    @score = case category
             when 'yacht' then dice.tally.one? ? 50 : 0
             when 'full house' then dice.tally.values.sort == [2, 3] ? dice.sum : 0
             when 'four of a kind' then dice.tally.find { _2 >= 4 }&.first.to_i * 4
             when 'little straight' then dice.sort == [1, 2, 3, 4, 5] ? 30 : 0
             when 'big straight' then dice.sort == [2, 3, 4, 5, 6] ? 30 : 0
             when 'choice' then dice.sum
             else number = %w[ones twos threes fours fives sixes].index(category) + 1
                  number * dice.tally.fetch(number, 0)
             end
  end
end
